freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java

```diff
@@ -45,89 +42,72 @@
-//            }
-            return getEnvToUmsStateMap(accountId, actorCrn, environmentsFilter, users, machineUsers, groups);
+            return getEnvToUmsStateMap(accountId, actorCrn, environmentsFilter, users, machineUsers);
 
         } catch (RuntimeException e) {
             throw new UmsOperationException(String.format("Error during UMS operation: %s", e.getMessage()));
         }
     }
 
     private Map<String, UmsState> getEnvToUmsStateMap(
-        String accountId, String actorCrn, Set<String> environmentsFilter, List<User> users, List<MachineUser> machineUsers, List<Group> groups) {
+        String accountId, String actorCrn, Set<String> environmentsFilter, List<User> users, List<MachineUser> machineUsers) {
         UmsState.Builder umsStateBuilder = new UmsState.Builder();
 
         // process each user and update environmentCRN -> UmsState map
         Map<String, UmsState> envUmsStateMap = new HashMap<>();
 
         environmentsFilter.stream().forEach(envCRN -> {
-            processForEnvironmentRights(umsStateBuilder, actorCrn, envCRN, users, machineUsers);
+            List<User> rightfulUsers = getUsersWithEnvironmentRights(actorCrn, envCRN, users);
+            umsStateBuilder.addUsers(rightfulUsers);
+
+            List<MachineUser> rightfulMachineUsers = getMachineUsersWithEnvironmentRights(actorCrn, envCRN, machineUsers);
+            umsStateBuilder.addMachineUsers(rightfulMachineUsers);
+
+            // get all groups for identified users those having rights.
+            Map<User, List<Group>> usersToGroupsMap = umsClient.getUsersToGroupsMap(actorCrn, accountId, rightfulUsers, Optional.empty());
+            umsStateBuilder.addUserToGroupMap(usersToGroupsMap);
+
+            Map<MachineUser, List<Group>> machineUsersToGroupsMap =
+                umsClient.getMachineUsersToGroupsMap(actorCrn, accountId, rightfulMachineUsers, Optional.empty());
+            umsStateBuilder.addMachineUserToGroupMap(machineUsersToGroupsMap);
+
+
             envUmsStateMap.put(envCRN, umsStateBuilder.build());
         });
 
-        // get all groups for users
-        Map<User, List<Group>> usersToGroupsMap = umsClient.getUsersToGroupsMap(actorCrn, accountId, users, Optional.empty());
-        umsStateBuilder.addUserToGroupMap(usersToGroupsMap);
-        //groups.stream().forEach(g -> umsStateBuilder.addGroup(g));
-
         return envUmsStateMap;
     }
 
-    private void processForEnvironmentRights(
-        Builder umsStateBuilder, String actorCrn, String envCRN, List<User> allUsers, List<MachineUser> allMachineUsers) {
+    private List<User> getUsersWithEnvironmentRights(
+        String actorCrn, String envCRN, List<User> allUsers) {
+        List<User> rightfulUsers = new ArrayList<>();
         // for all users, check right for the passed envCRN
         for (User u : allUsers) {
 
             if (umsClient.checkRight(actorCrn, u.getCrn(), environmentWrite, envCRN, Optional.empty())) {
```

Comment:
does this work? environments/write is resourceIndependent just like setPassword is and should hit the same illegal argument error

Reply:
This is an Invalid Argument I just commented it since this is already there and this is not for master branch as of now.